lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -3544,13 +3551,21 @@ CheckedCastKind TypeChecker::typeCheckCheckedCast(Type fromType,
     // Note: we relax the restriction if the type we're casting to is a
     // non-final class because it's possible that we might have a subclass
     // that conforms to the protocol.
+    //
+    //  Also, relax the restriction when toType is a stdlib collection
+    //  (Array, Set, Dictionary) that could have conditional conformances
+    //  because of custom casting mechanism implemented for those types e.g.
+    //
+    //  func encodable(_ value: Encodable) {
+    //    _ = value as! [String : Encodable]
+    //  }
+    //
     if (fromExistential && !toExistential) {
       if (auto NTD = toType->getAnyNominal()) {
         if (!toType->is<ClassType>() || NTD->isFinal()) {
```

Comment:
I think this check should be done in `couldDynamicallyConformToProtocol`, and could be done as an early return, something along the lines of:

```cxx
// If this is a non-final class type, its subclass could potentially conform.
if (auto *classDecl = type->getClassOrBoundGenericClass()) {
  if (!classDecl->isFinal())
    return true;
}
```

Reply:
Ahh now I see, thanks I'll adjust that :)